src/core/main.c

```diff
@@ -1144,20 +1144,35 @@ static int prepare_reexecute(
         if (!fds)
                 return log_oom();
 
+        /* We need existing BPF programs to survive reload, otherwise there will be a period where no BPF
+         * program is active during task execution within a cgroup. This would be bad since this may have
+         * security or reliability implications: devices we should filter won't be filtered, network activity
+         * we should filter won't be filtered, etc. We pin all the existing devices by bumping their
+         * refcount, and then storing them to later have it decremented. */
+        (void) manager_pin_all_cgroup_bpf_programs(m);
+
         r = manager_serialize(m, f, fds, switching_root);
-        if (r < 0)
+        if (r < 0) {
+                manager_unpin_all_cgroup_bpf_programs(m);
```

Comment:
Let me know if you'd prefer this as a cleanup goto label. I lightly prefer this way, since it's not too unwieldy.

Reply:
It's pretty ugly without automatic cleanup ;(

Reply:
Hmm, what would you recommend? I don't think we can use cleanup attribute here, since if it succeeds we actually need to keep it alive, unless I'm missing something?

Another option is to TAKE_PTR it all the way back to `main()`, but that seems even more ugly.

Reply:
Yeah, normally we'd put `TAKE_PTR(m)` at the point where the function is known to succeed. But current form is OK too, it doesn't matter too much.